dask/dataframe/io/orc.py

```diff
@@ -0,0 +1,78 @@
+from __future__ import absolute_import, division, print_function
+
+from ..core import DataFrame
+from ...base import tokenize
+from ...bytes.core import get_fs_token_paths
+from ...utils import import_required
+from .parquet import _get_pyarrow_dtypes, _meta_from_dtypes
+from ...compatibility import string_types
+
+__all__ = ('read_orc',)
+
+
+def _read_orc_stripe(fs, path, stripe, columns=None):
+    """Pull out specific data from specific part of ORC file"""
+    orc = import_required('pyarrow.orc', 'Please install dev version pyarrow')
+    with fs.open(path, 'rb') as f:
+        o = orc.ORCFile(f)
+        return o.read_stripe(stripe, columns).to_pandas()
+
+
+def read_orc(path, columns=None, storage_options=None):
+    """Read dataframe from ORC file(s)
+
+    Parameters
+    ----------
+    path: str or list(str)
+        Location of file(s), which can be a full URL with protocol specifier,
+        and may include glob character if a single string.
+    columns: None or list(str)
+        Columns to load. If None, loads all.
+    storage_options: None or dict
+        Further parameters to pass to the bytes backend.
+
+    Returns
+    -------
+    Dask.DataFrame (even if there is only one column)
+
+    Examples
+    --------
+    >>> d = read_orc('https://github.com/apache/orc/raw/'
+    ...              'master/examples/demo-11-zlib.orc')  # doctest: +SKIP
+    """
+    orc = import_required('pyarrow.orc', 'Please install dev version pyarrow')
+    if isinstance(path, string_types) and path.endswith('/'):
+        path = path + '*.orc'
+    storage_options = storage_options or {}
+    fs, fs_token, paths = get_fs_token_paths(path, mode='rb',
+                                             storage_options=storage_options)
+    schema = None
+    nstripes = []
+    for path in paths:
+        with fs.open(path, 'rb') as f:
+            o = orc.ORCFile(f)
+            if schema is None:
+                schema = o.schema
+            elif schema != o.schema:
+                raise ValueError("Incompatible schemas while parsing ORC files")
```

Comment:
I'm not actually sure what the best method is here. This is slightly inefficient, but lets us error at build time rather than runtime. For now this seems fine to me.

Reply:
We need to at least know the number of stripes, as it becomes the number of partitions, so there's probably no getting around it.